examples/dtls-echo/dtls-client.c

```diff
@@ -210,325 +199,360 @@ static int peer_get_ecdsa_key(struct dtls_context_t *ctx,
+            remote->netif = (uint16_t)gnrc_netif_iter(NULL)->pid;
+        }
+        else {
+            /* FIXME This probably is not valid with multiple interfaces */
+            dst->ifindex = remote->netif;
+        }
+    }
+    else {
+        if (gnrc_netif_get_by_pid(iface) == NULL) {
+            puts("ERROR: interface not valid");
+            return new_context;
+        }
+        dst->ifindex = (uint16_t)gnrc_netif_iter(NULL)->pid;
+        remote->netif = (uint16_t)gnrc_netif_iter(NULL)->pid;
     }
 
-    ipv6_addr_t addr_dbg;
-    ipv6_addr_from_str(&addr_dbg, addr_str);
+    if (ipv6_addr_from_str((ipv6_addr_t *)remote->addr.ipv6, addr_str) == NULL) {
+        puts("ERROR: unable to parse destination address");
+        return new_context;
+    }
 
-    /*akin to syslog: EMERG, ALERT, CRITC, NOTICE, INFO, DEBUG */
-    dtls_set_log_level(DTLS_LOG_NOTICE);
+    /* Second: We prepare the DTLS Session by means of ctx->app */
+    dst->size = sizeof(uint8_t) * 16 + sizeof(unsigned short);
+    dst->port = remote->port;
 
-    dtls_context = dtls_new_context(addr_str);
-    if (dtls_context) {
-        dtls_set_handler(dtls_context, &cb);
+    /* NOTE: remote.addr.ipv6 and dst->addr are different structures. */
+    if (ipv6_addr_from_str(&dst->addr, addr_str) == NULL) {
+        puts("ERROR: init_dtls was unable to load the IPv6 addresses!");
+        return new_context;
     }
 
-    return;
+    new_context = dtls_new_context(sock);
+    if (new_context) {
+        dtls_set_handler(new_context, &cb);
+    }
+
+    return new_context;
 }
 
-/**
- * This is the "client" part of this program.
- * Will be called each time a message is transmitted.
- */
-static void client_send(char *addr_str, char *data, unsigned int delay)
+static void client_send(char *addr_str, char *data,
+                        unsigned int delay)
 {
-    int8_t iWatch;
     static session_t dst;
-    static int connected = 0;
-    msg_t msg;
+    dtls_context_t *dtls_context = NULL;
+
+    sock_udp_ep_t local = SOCK_IPV6_EP_ANY;
+    sock_udp_ep_t remote = SOCK_IPV6_EP_ANY;
+    sock_udp_t sock;
+
+    uint8_t packet_rcvd[DTLS_MAX_BUF];
+    int8_t watch = MAX_TIMES_TRY_TO_SEND;
+
+    ssize_t pckt_rcvd_size = DTLS_MAX_BUF;  /* DTLS flights records */
+    ssize_t app_data_buf = 0;               /* Upper layer packet to send */
 
-   gnrc_netreg_entry_t entry = GNRC_NETREG_ENTRY_INIT_PID(CLIENT_PORT,
-                                                           sched_active_pid);
-    dtls_init();
 
-    if (gnrc_netreg_register(GNRC_NETTYPE_UDP, &entry)) {
-        puts("Unable to register ports");
-        /*FIXME: Release memory?*/
+    /* NOTE: dtls_init() must be called previous to this (see main.c) */
+
+    dtls_context = _init_dtls(&sock, &local, &remote, &dst, addr_str);
+    if (!dtls_context) {
+        puts("ERROR: Client unable to load context!");
         return;
     }
 
+#ifndef COAP_MSG_SPOOF
```

Comment:
This defines seems unrelated. Can you explain it ?

Reply:
Ah, true. I put the comment in the Makefile, but not in the README.md neither here. When we were working with pyaiot + dtls and OTA, I used this example for testing my CoAPS nodes. And to be honest, I found it very useful.

In brief,  `COAP_MSG_SPOOF` changes the behavior of the client. When is settled the client sends a "CoAP get /riot/value" message instead of a user input. It's also changing the output format of incoming messages by treating them like raw bytes values instead of ASCII.